### PR TITLE
[Core][Pubsub][Logging 2/n] Use GCS pubsub for logs

### PR DESCRIPTION
**Pull request by mwtian**
## Why are these changes needed?
Using Ray pubsub for publishing and subscribing logs via GCS, from Python worker, log importer, dashboard and unit tests.

This change is guarded behind the `RAY_gcs_grpc_based_pubsub` feature flag.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: src/ray/protobuf/common.proto
Comment (mwtian) [inline src/ray/protobuf/common.proto:557]: Feel free to suggest alternative files for this.

File: src/ray/gcs/gcs_server/gcs_server.cc
Comment (iycheng) [inline src/ray/gcs/gcs_server/gcs_server.cc:78]: I'm surprised that it's working in CPP. I mean there is a tailing ','😀

File: python/ray/_private/gcs_pubsub.py
Comment (iycheng) [inline python/ray/_private/gcs_pubsub.py:206]: Please add a type hint for this field to help future maintainers.

Comment (iycheng) [reply python/ray/_private/gcs_pubsub.py:206]: Or just comment of this one. It's hard to tell what's this for just from here.

File: python/ray/_private/gcs_pubsub.py
Comment (iycheng) [inline python/ray/_private/gcs_pubsub.py:205]: It'll also be useful to add some inline comments about what are these protecting

File: python/ray/_private/gcs_pubsub.py
Comment (iycheng) [inline python/ray/_private/gcs_pubsub.py:301]: The comment is wrong? Do we throw an exception here?

**Review by ericl — CHANGES_REQUESTED** [review]
This PR is pretty big. Could we split it into a couple pieces? Seems like we can first refactor the python code to make the pubsub impl pluggable, then separately add a GCS based implementation.

File: python/ray/_private/gcs_pubsub.py
Comment (iycheng) [inline python/ray/_private/gcs_pubsub.py:76]: Add type hint here and comment on the usage for this field.

**Review by fishbone — COMMENTED** [review]
Thanks for this work! It looks like right now all threads are going to the same code path, including polling. I feel the code is a little bit complicated when read _poll.

Will it be better if we have an internal thread in subscriber and it's doing polling. It'll push the result to the task queue which is thread safe. The subscribe will just poll from the specific queue. will this be better?

File: python/ray/_private/gcs_pubsub.py
Comment (mwtian) [reply python/ray/_private/gcs_pubsub.py:301]: Reworded to make it clear that it is intended that `fut.result()` throws exception on failure.

**Comment by mwtian** [comment]
> This PR is pretty big. Could we split it into a couple pieces? Seems like we can first refactor the python code to make the pubsub impl pluggable, then separately add a GCS based implementation.

Sent out a refactoring change without GCS pubsub implementation in #20560.

> Will it be better if we have an internal thread in subscriber and it's doing polling. It'll push the result to the task queue which is thread safe. The subscribe will just poll from the specific queue. will this be better?

Sounds good. I was hoping to save a thread and make error propagation easier but it may not worth the complexity.

**Comment by mwtian** [comment]
Will merge master with #20560 into this PR after #20560 is merged into master.

File: src/ray/protobuf/common.proto
Comment (raulchen) [reply src/ray/protobuf/common.proto:559]: Will this be used in something other than PubSub? If not, putting it in `pubsub.proto` may be better.

File: src/ray/gcs/gcs_server/gcs_server.cc
Comment (raulchen) [reply src/ray/gcs/gcs_server/gcs_server.cc:78]: This might be something new in C++17? I remember this didn't work before. 
I have been wanting this for long time. It would mitigate our pains of handling code conflicts.

File: python/ray/_private/gcs_pubsub.py
Comment (raulchen) [inline python/ray/_private/gcs_pubsub.py:74]: Core worker already has a PubSub client. Why not use that one? 
Can reduce the number of connections to GCS.

Comment (mwtian) [reply python/ray/_private/gcs_pubsub.py:74]: Python publisher and subscribers are necessary for autoscaler and dashboard, which cannot depend on the worker library. @iycheng and I have discussed reusing gcs channel in core worker for Python workers. Because of the additional work needed to implement this optimization, we decided to just use a common GCS channel in Python pubsub and KV clients. In future we can implement an optimization to communicate to GCS via raylet.

**Review by mwtian — COMMENTED** [review]
The Python pubsub refactoring and logging feature are reviewed in https://github.com/ray-project/ray/pull/20604.

File: src/ray/gcs/gcs_server/gcs_server.cc
Comment (mwtian) [reply src/ray/gcs/gcs_server/gcs_server.cc]: Yeah, not sure when it became available!

**Comment by mwtian** [comment]
Rebased on top of #20604 which added Python supporting for log publishing and subscribing. Previous comments should have been addressed. This change is guarded behind the `RAY_gcs_grpc_based_pubsub` feature flag.

**Review by fishbone — APPROVED** [review]
Looks clean and thanks!

**Comment by iycheng** [comment]
@ericl  has requested the changes, I'll let him go with a final review.

**Review by ericl — APPROVED** [review]

**Pull request merged**
